### PR TITLE
Draft PR for Question 1

### PR DESCRIPTION
**Pull request by inf0rmatiker**
## Changes

- Added `schemas` package with schema structs for both `DataFrame` types.
- Added `loading` package for loading in both *.txt files from HDFS into `DataFrame`s, using aforementioned schemas.
- Begin implementation of first question as part of `analytics` package.

### CODE REVIEW

File: src/main/scala/org/citegraph/Application.scala [hunk 1]
```diff
@@ -13,18 +20,51 @@ object Application {
...
+    // Validate and fix inputDirectory
+    var inputDirectory: String = args(0)
+    val outputDirectory: String = args(1)
+    if (!isValidHdfsUri(inputDirectory)) {
+      printf("Invalid HDFS input directory: %s\n", inputDirectory)
+      System.exit(1)
+    } else if (!isValidHdfsUri(outputDirectory)) {
+      printf("Invalid HDFS output directory: %s\n", outputDirectory)
+      System.exit(1)
+    }
```
Comment (NFabrizio) [inline src/main/scala/org/citegraph/Application.scala:45]: These guard statements probably aren't necessary, but I like having them in here.

File: src/main/scala/org/citegraph/Application.scala [hunk 2]
```diff
@@ -13,18 +20,51 @@ object Application {
...
+
+
+    val dataframeLoader: DataFrameLoader = new DataFrameLoader(inputDirectory, sparkSession)
+    val citationsDF: DataFrame = dataframeLoader.loadCitations()
+    val publishedDatesDF: DataFrame = dataframeLoader.loadPublishedDates()
+
+    val analytics: Analytics = new Analytics(sparkSession, citationsDF, publishedDatesDF)
+    analytics.findDensitiesByYear()
+
+    sparkSession.close()
```
Comment (NFabrizio) [inline src/main/scala/org/citegraph/Application.scala:70]: This is very clean, concise and easy to read. Nicely done.

File: src/main/scala/org/citegraph/analytics/Analytics.scala [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      +----+-----+
+      |1997| 4252|
+      |1994| 2894|
+      |1996| 3939|
+      |... |...  |
+      +----+-----+
+     */
+    publishedDatesDF.groupBy(col("year"))
+      .count()
+      .show()
```
Comment (NFabrizio) [inline src/main/scala/org/citegraph/analytics/Analytics.scala]: I think we can probably get this to print in the same format as the template in the assignment description, but this is good for now.

File: src/main/scala/org/citegraph/loading/DataFrameLoader.scala [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+      |9907233|9506257|
+      |9907233|9606402|
+      |9907233|9607354|
+      |...    |...    |
+      +-------+-------+
+   */
+  def loadCitations(): DataFrame = {
+    val sparkContext: SparkContext = sparkSession.sparkContext
+    val citationsRDD: RDD[Row] = sparkContext.textFile(s"$dataDirectory/$CITATIONS_TXT_STR")
+      .filter(line => !line.contains("#") && line.trim().nonEmpty) // Remove lines that contain '#' and empty lines
```
Comment (NFabrizio) [inline src/main/scala/org/citegraph/loading/DataFrameLoader.scala:31]: Removing empty lines is nice.

File: src/main/scala/org/citegraph/loading/DataFrameLoader.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+      .map(line => {
+        val lineParts: Array[String] = line.split("\\s+") // Split on whitespace
+        Row(lineParts(0).trim().toInt, lineParts(1).trim().toInt) // Output Row[Integer, Integer]
```
Comment (NFabrizio) [inline src/main/scala/org/citegraph/loading/DataFrameLoader.scala:34]: Good idea casting these to ints.

**Review by NFabrizio — APPROVED** [review]

File: src/main/scala/org/citegraph/Application.scala (hunk 2 above)
Comment (inf0rmatiker) [reply src/main/scala/org/citegraph/Application.scala:70]: Thanks!

File: src/main/scala/org/citegraph/analytics/Analytics.scala (hunk 3 above)
Comment (inf0rmatiker) [reply src/main/scala/org/citegraph/analytics/Analytics.scala]: Noted; I've added a .csv saver called `DataFrameSaver` under pkg `saving`. It coalesces to 1 worker, can sort on a column (in this case, "year"), and saves to HDFS output dir you specify in program args.

That 'findDensitiesByYear()` function is not complete; we'll need to add the other two columns and name them appropriately

**Pull request merged**